api/grpc/controller_test.go

```diff
@@ -0,0 +1,263 @@
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package grpc
+
+import (
+	"context"
+	"errors"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	tmock "github.com/stretchr/testify/mock"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/model/flow"
+
+	"github.com/optakt/flow-dps/models/dps/mock"
+)
+
+func TestNewController(t *testing.T) {
+	c := NewController(nil)
+	assert.NotNil(t, c)
+}
+
+func TestController_GetRegister(t *testing.T) {
+	var (
+		testHeight uint64 = 128
+		lastHeight uint64 = 256
+
+		testKey   = []byte(`testKey`)
+		testValue = []byte(`testValue`)
+	)
+
+	tests := []struct {
+		desc string
+
+		reqHeight *uint64
+		reqKey    []byte
+
+		mockValue []byte
+		mockErr   error
+
+		wantHeight uint64
+		wantResp   *GetRegisterResponse
+		wantErr    assert.ErrorAssertionFunc
+	}{
+		{
+			desc: "nominal case, height given",
+
+			reqHeight: &testHeight,
+			reqKey:    testKey,
+
+			mockValue: testValue,
+
+			wantHeight: testHeight,
+			wantResp: &GetRegisterResponse{
+				Height: testHeight,
+				Key:    testKey,
+				Value:  testValue,
+			},
+			wantErr: assert.NoError,
+		},
+		{
+			desc: "nominal case, no height given",
+
+			reqKey: testKey,
+
+			mockValue: testValue,
+
+			wantHeight: lastHeight,
+			wantResp: &GetRegisterResponse{
+				Height: lastHeight,
+				Key:    testKey,
+				Value:  testValue,
+			},
+			wantErr: assert.NoError,
+		},
+		{
+			desc: "state error",
+
+			reqKey: testKey,
+
+			mockErr: errors.New("dummy error"),
+
+			wantHeight: lastHeight,
+			wantResp:   nil,
+			wantErr:    assert.Error,
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+		t.Run(test.desc, func(t *testing.T) {
+			t.Parallel()
+
+			m := mock.NewState()
+			m.LastState.On("Height").Return(lastHeight).Once()
+			m.RawState.On("WithHeight", test.wantHeight).Return(m.RawState).Once()
+			m.RawState.On("Get", test.reqKey).Return(test.mockValue, test.mockErr)
+
+			c := &Controller{
+				state: m,
+			}
+
+			req := &GetRegisterRequest{
+				Height: test.reqHeight,
+				Key:    test.reqKey,
+			}
+
+			got, err := c.GetRegister(context.Background(), req)
+			test.wantErr(t, err)
+
+			if test.wantResp != nil {
+				assert.Equal(t, test.wantResp, got)
+			}
+
+			m.AssertExpectations(t)
+		})
+	}
+}
+
+func TestController_GetValues(t *testing.T) {
+	var (
+		testKeys = []*Key{
+			{
+				Parts: []*KeyPart{
+					{
+						Type:  0,
+						Value: []byte(`testOwner`),
+					},
+					{
+						Type:  1,
+						Value: []byte(`testController`),
+					},
+					{
+						Type:  2,
+						Value: []byte(`testKey`),
+					},
+				},
+			},
+		}
+		testVersion uint64 = 42
```

Comment:
This is like a super nitpick, but if we use type inference for all others, we could also use it here with `testVersion = uint64(42)`. Looks strange to only declare the type there.